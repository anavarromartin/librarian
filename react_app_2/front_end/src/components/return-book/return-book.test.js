import React from 'react';
import renderer from 'react-test-renderer';
import ReturnBook from "./return-book"
import {mount, shallow} from 'enzyme'
import 'jasmine-enzyme'

const mountComponent = ({
                            setBackLocation = () => {
                            },
                            setHeaderVisibility = () => {
                            },
                            getCheckedOutBooks = () => {
                            }
                        } = {}) => (
    mount(<ReturnBook
        setBackLocation={setBackLocation}
        setHeaderVisibility={setHeaderVisibility}
        getCheckedOutBooks={getCheckedOutBooks}
    />)
)

const clickOnSearchInput = component => {
    component.find('.labels-input__input').simulate('click')
}

const typeOnSearchInput = (text, component) => {
    component.find(".labels-input__input").simulate('change', {target: {value: text}})
}

describe('<ReturnBook />', () => {

    describe('display', () => {
        it('renders the component', () => {
            let component
                component = renderer.create(<ReturnBook setBackLocation={() => {
                }}/>)
            expect(component.toJSON()).toMatchSnapshot()
        })

        describe('when clicking on search input', () => {
            let component
            beforeEach(() => {
                component = mountComponent()
                clickOnSearchInput(component)
            })

            it('renders search mode', () => {
                expect(component).toMatchSnapshot()
            })

            describe('when searching', () => {
                it('renders search results', async () => {
                    const booksPromise = Promise.resolve(
                        [
                            {
                                id: 1,
                                name: 'a book',
                                checkout_histories: [
                                    {
                                        name: 'adria',
                                    }
                                ]
                            }
                        ])

                    component = mountComponent({
                        getCheckedOutBooks: () => {
                            return booksPromise
                        }
                    })

                    typeOnSearchInput('a book', component)
                    await booksPromise

                    expect(component).toMatchSnapshot()
                })
            })
        })
    })

    describe('behavior', () => {
        it('when typing it performs search', () => {
            let searchInput = ''

            typeOnSearchInput('Test', mountComponent({
                getCheckedOutBooks: search => {
                    searchInput = search
                }
            }))

            expect(searchInput).toEqual('Test')
        })

        it('when input is empty it does not search', () => {
            let wasCalled = false

            typeOnSearchInput('   ', mountComponent({
                getCheckedOutBooks: () => {
                    wasCalled = true
                }
            }))

            expect(wasCalled).toBeFalsy()
        })
    })
})