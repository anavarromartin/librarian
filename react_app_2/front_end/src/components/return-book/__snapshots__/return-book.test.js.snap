// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReturnBook /> display renders the component 1`] = `
<div
  className="container"
>
  <div
    className="header"
  >
    Return A Book
  </div>
  <div
    className="form"
  >
    <label
      className="label"
    >
      Find the Book You Borrowed
    </label>
    <div
      className=""
    >
      <div
        className="search__field"
      >
        <div
          className="search__back"
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-chevron-left fa-w-10 search__back-icon"
            data-icon="chevron-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
        <div
          className="search__input"
        >
          <div
            className="label-input__container"
          >
            <div
              className="labels__container"
              style={
                Object {
                  "display": "none",
                }
              }
            />
            <input
              className="labels-input__input"
              onChange={[Function]}
              onClick={[Function]}
              style={
                Object {
                  "display": "block",
                }
              }
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="search__results"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          className="search-results__top-container"
        />
      </div>
    </div>
    <button
      className="teal-button button-input"
      onClick={[Function]}
    >
      RETURN
    </button>
  </div>
</div>
`;

exports[`<ReturnBook /> display when clicking on search input renders search mode 1`] = `
<ReturnBook
  getCheckedOutBooks={[Function]}
  history={
    Object {
      "push": [Function],
    }
  }
  returnBook={[Function]}
  setBackLocation={[Function]}
  setHeaderVisibility={[Function]}
>
  <div
    className="container"
  >
    <div
      className="header"
    >
      Return A Book
    </div>
    <div
      className="form"
    >
      <label
        className="label"
      >
        Find the Book You Borrowed
      </label>
      <SearchInput
        isSearching={[Function]}
        onSelectResult={[Function]}
        search={[Function]}
      >
        <div
          className="searching"
        >
          <div
            className="search__field"
          >
            <div
              className="search__back"
              onClick={[Function]}
            >
              <FontAwesomeIcon
                border={false}
                className="search__back-icon"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      320,
                      512,
                      Array [],
                      "f053",
                      "M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z",
                    ],
                    "iconName": "chevron-left",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-chevron-left fa-w-10 search__back-icon"
                  data-icon="chevron-left"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 320 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </div>
            <div
              className="search__input"
            >
              <LabelInput
                clearLabel={[Function]}
                labels={Array []}
                onChange={[Function]}
                onClick={[Function]}
                value=""
              >
                <div
                  className="label-input__container"
                >
                  <div
                    className="labels__container"
                    style={
                      Object {
                        "display": "none",
                      }
                    }
                  />
                  <input
                    className="labels-input__input"
                    onChange={[Function]}
                    onClick={[Function]}
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                    type="text"
                    value=""
                  />
                </div>
              </LabelInput>
            </div>
          </div>
          <div
            className="search__results"
            style={
              Object {
                "display": "none",
              }
            }
          >
            <SearchResults
              onSelectResult={[Function]}
              results={Array []}
            >
              <div
                className="search-results__top-container"
              />
            </SearchResults>
          </div>
        </div>
      </SearchInput>
      <button
        className="teal-button button-input"
        onClick={[Function]}
      >
        RETURN
      </button>
    </div>
  </div>
</ReturnBook>
`;

exports[`<ReturnBook /> display when clicking on search input when searching renders search results 1`] = `
<ReturnBook
  getCheckedOutBooks={[Function]}
  history={
    Object {
      "push": [Function],
    }
  }
  returnBook={[Function]}
  setBackLocation={[Function]}
  setHeaderVisibility={[Function]}
>
  <div
    className="container"
  >
    <div
      className="header"
    >
      Return A Book
    </div>
    <div
      className="form"
    >
      <label
        className="label"
      >
        Find the Book You Borrowed
      </label>
      <SearchInput
        isSearching={[Function]}
        onSelectResult={[Function]}
        search={[Function]}
      >
        <div
          className=""
        >
          <div
            className="search__field"
          >
            <div
              className="search__back"
              onClick={[Function]}
            >
              <FontAwesomeIcon
                border={false}
                className="search__back-icon"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      320,
                      512,
                      Array [],
                      "f053",
                      "M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z",
                    ],
                    "iconName": "chevron-left",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-chevron-left fa-w-10 search__back-icon"
                  data-icon="chevron-left"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 320 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </div>
            <div
              className="search__input"
            >
              <LabelInput
                clearLabel={[Function]}
                labels={Array []}
                onChange={[Function]}
                onClick={[Function]}
                value="a book"
              >
                <div
                  className="label-input__container"
                >
                  <div
                    className="labels__container"
                    style={
                      Object {
                        "display": "none",
                      }
                    }
                  />
                  <input
                    className="labels-input__input"
                    onChange={[Function]}
                    onClick={[Function]}
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                    type="text"
                    value="a book"
                  />
                </div>
              </LabelInput>
            </div>
          </div>
          <div
            className="search__results"
            style={
              Object {
                "display": "none",
              }
            }
          >
            <SearchResults
              onSelectResult={[Function]}
              results={Array []}
            >
              <div
                className="search-results__top-container"
              />
            </SearchResults>
          </div>
        </div>
      </SearchInput>
      <button
        className="teal-button button-input"
        onClick={[Function]}
      >
        RETURN
      </button>
    </div>
  </div>
</ReturnBook>
`;

exports[`<ReturnBook /> display when clicking on search input when searching when selecting book renders selected result 1`] = `
<ReturnBook
  getCheckedOutBooks={[Function]}
  history={
    Object {
      "push": [Function],
    }
  }
  returnBook={[Function]}
  setBackLocation={[Function]}
  setHeaderVisibility={[Function]}
>
  <div
    className="container"
  >
    <div
      className="header"
    >
      Return A Book
    </div>
    <div
      className="form"
    >
      <label
        className="label"
      >
        Find the Book You Borrowed
      </label>
      <SearchInput
        isSearching={[Function]}
        onSelectResult={[Function]}
        search={[Function]}
      >
        <div
          className=""
        >
          <div
            className="search__field"
          >
            <div
              className="search__back"
              onClick={[Function]}
            >
              <FontAwesomeIcon
                border={false}
                className="search__back-icon"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      320,
                      512,
                      Array [],
                      "f053",
                      "M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z",
                    ],
                    "iconName": "chevron-left",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-chevron-left fa-w-10 search__back-icon"
                  data-icon="chevron-left"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 320 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </div>
            <div
              className="search__input"
            >
              <LabelInput
                clearLabel={[Function]}
                labels={
                  Array [
                    Object {
                      "book_name": "a book",
                      "borrower_name": "adria",
                      "id": 1,
                    },
                  ]
                }
                onChange={[Function]}
                onClick={[Function]}
                value=""
              >
                <div
                  className="label-input__container"
                >
                  <div
                    className="labels__container"
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                  >
                    <div
                      className="labels-input__label"
                      key="1"
                    >
                      <div
                        className="labels-input__label-text"
                      >
                        a book
                         - adria
                      </div>
                      <FontAwesomeIcon
                        border={false}
                        className="labels-input__label-remove"
                        fixedWidth={false}
                        flip={null}
                        icon={
                          Object {
                            "icon": Array [
                              352,
                              512,
                              Array [],
                              "f00d",
                              "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
                            ],
                            "iconName": "times",
                            "prefix": "fas",
                          }
                        }
                        inverse={false}
                        listItem={false}
                        mask={null}
                        onClick={[Function]}
                        pull={null}
                        pulse={false}
                        rotation={null}
                        size={null}
                        spin={false}
                        symbol={false}
                        title=""
                        transform={null}
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-times fa-w-11 labels-input__label-remove"
                          data-icon="times"
                          data-prefix="fas"
                          focusable="false"
                          onClick={[Function]}
                          role="img"
                          style={Object {}}
                          viewBox="0 0 352 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </FontAwesomeIcon>
                    </div>
                  </div>
                  <input
                    className="labels-input__input"
                    onChange={[Function]}
                    onClick={[Function]}
                    style={
                      Object {
                        "display": "none",
                      }
                    }
                    type="text"
                    value=""
                  />
                </div>
              </LabelInput>
            </div>
          </div>
          <div
            className="search__results"
            style={
              Object {
                "display": "none",
              }
            }
          >
            <SearchResults
              onSelectResult={[Function]}
              results={Array []}
            >
              <div
                className="search-results__top-container"
              />
            </SearchResults>
          </div>
        </div>
      </SearchInput>
      <button
        className="teal-button button-input"
        onClick={[Function]}
      >
        RETURN
      </button>
    </div>
  </div>
</ReturnBook>
`;
